### ISSUE TITLE
Memory issue with offscreen rendering

### ISSUE THREAD

**Issue by danburzo** [electron/electron#7350]
- Electron version: 1.4.1
- Operating system: macOS 10.12 (Sierra)

When using offscreen rendering, the memory used by this grows uncontrollably:

``` js
var app = require('electron').app;
var BrowserWindow = require('electron').BrowserWindow;

app.disableHardwareAcceleration();

var win, iteration = 0;

function leakke_li() {
    if (win) {
        win.destroy();
    }
    win = new BrowserWindow({
        width: 2000,
        height: 2000,
        webPreferences: {
            offscreen: true
        },
        show: false
    });
    win.webContents.on('paint', function(event, dirty, image) {
        console.log('event:paint');
    });
    win.webContents.on('did-finish-load', function() {
        console.log('event:did-finish-load', ++iteration);
        setTimeout(leakke_li, 300);
    });
    win.loadURL('about:blank');
}

app.on('ready', function() {
    var main = new BrowserWindow({ width: 10, height: 10 });
    leakke_li();
});
```

**Note:** When using `offscreen:  false`, memory stays mostly constant.

Any idea why this might happen?

**Comment by danburzo** [comment]
I've tested the offscreen version on Electron 1.4.0 and it seems the memory stabilizes at ~2.5GB, but after this limit is reached the performance drops severely.

**Comment by gerhardberger** [comment]
This issue is only present in the GPU-accelerated offscreen rendering mode.

In electron 1.4.0-1 the software output mode isn't working, that's why it switches back to GPU mode, but I've already submitted a fix for it, #7327, hopefully it will be available in the next release.

I don't know that this issue is present on Windows, but after a little investigation I see that, after the bitmap is [transformed into `NativeImage`](https://github.com/electron/electron/blob/master/atom/browser/api/atom_api_web_contents.cc#L1368) and the webContents is destroyed the image stays in memory. When I don't create a `NativeImage` from it, or I call an `image.Clear();` directly after the `"paint"` emitting, it goes out of memory.

**Comment by danburzo** [comment]
Hi @gerhardberger, I had seen your pull request but I think I misread it the first time around and didn't link it to my problem. To paraphrase it just to make sure I got this right: sounds like `disableHardwareAcceleration` is mostly placebo until that gets merged, and the memory spiraling out of control should be fixed once there's true software rendering?

**Comment by gerhardberger** [comment]
@danburzo, yes currently calling `disableHardwareAcceleration` has no effect, and the memory problem won't be present in the software output mode, when it will work again.

**Comment by danburzo** [comment]
Music to my ears 💃

**Comment by danburzo** [comment]
I've rerun the test same test (but set the width/height to 10000 pixels) with Electron v.1.4.2 on macOS Sierra and here's what I got:

1. Memory usage is _much, much better_, but still grows from one iteration to another
2. At about iteration 100, the rhythm slows down considerably
3. It finally crashes with:

```bash
Electron(9093,0x7fffcc34f3c0) malloc: *** error for object 0x7fc3276efe10: pointer being freed was not allocated
*** set a breakpoint in malloc_error_break to debug
./start.sh: line 3:  9093 Abort trap: 6           bin/Electron.app/Contents/MacOS/Electron ./index.js $*
```

This also seems to be the case for a more decent-sized, 2000 by 2000 pixels, window.

**Comment by danburzo** [comment]
The memory leak seems to have been fully eradicated in [Electron 1.4.3](https://github.com/electron/electron/releases/tag/v1.4.3) most likely with [this pull request](https://github.com/electron/electron/pull/7452) (thanks, @enlight! 🎉).

Still crashes eventually, (and there's the inevitable slowdown after several iterations), but it's in much better shape.

**Comment by danburzo** [comment]
__However__ neither version between Electron 1.4.0-1.4.3 crashes on a headless Ubuntu 14 (using Electron with xvfb), but rather freezes and a weird child process appears. Here's `htop` after it froze:
<img width="1270" alt="screen shot 2016-10-07 at 3 50 16 pm" src="https://cloud.githubusercontent.com/assets/205375/19190457/43c14c8e-8ca6-11e6-83df-de9a417ed9c7.png">

@gerhardberger any idea how we could debug this? Or a reliable way to detect that it froze?

**Comment by amoshg** [comment]
@danburzo did you end up finding the root cause or a soltuon for the freezing when using Electron with xvfb?

**Comment by danburzo** [comment]
@amoshg unfortunately we have not been able to reliably prevent the freezing, so we've worked around it by refreshing the electron processes periodically, and having fallbacks to different processes when we detect an abnormal timeout (which most likely means it froze)

**Comment by alecperkey** [comment]
I don't know if this is useful but here's some tests I did

https://bugs.chromium.org/p/chromium/issues/detail?id=805655
https://github.com/Microsoft/vscode/issues/42122

**Comment by brenca** [comment]
I'm closing this, since the original issue was solved, and electron/native-mate#12 should have improved this further.
